reward/supply_manager.go

```diff
@@ -0,0 +1,377 @@
+	errNoAccReward       = errors.New("accumulated reward not stored")
+	errNoBlock           = errors.New("block not found")
+	errNoRebalanceMemo   = errors.New("rebalance memo not yet stored")
+)
+
+// SupplyManager tracks the total supply of native tokens.
+// Note that SupplyManager only deals with the block rewards.
+// Factors like KIP103, KIP160, 0xdead burning are not considered and must be accounted
+// in the APIs that return the total supply.
+type SupplyManager interface {
+	// Start starts the supply manager goroutine that catches up or tracks the block rewards.
+	Start()
+
+	// Stop stops the supply manager goroutine.
+	Stop()
+
+	// GetTotalSupply returns the total supply amounts at the given block number,
+	// broken down by minted amount and burnt amounts of each methods.
+	GetTotalSupply(num uint64) (*TotalSupply, error)
+}
+
+type TotalSupply struct {
+	TotalSupply *big.Int // The total supply of the native token. i.e. Minted - Burnt.
+	TotalMinted *big.Int // Total minted amount.
+	TotalBurnt  *big.Int // Total burnt amount. Sum of all burnt amounts below.
+	BurntFee    *big.Int // from tx fee burn. ReadAccReward(num).BurntFee.
+	ZeroBurn    *big.Int // balance of 0x0 (zero) address.
+	DeadBurn    *big.Int // balance of 0xdead (dead) address.
+	Kip103Burn  *big.Int // by KIP103 fork. Read from its memo.
+	Kip160Burn  *big.Int // by KIP160 fork. Read from its memo.
+}
+
+type supplyManager struct {
+	// Externally injected dependencies
+	chain              blockChain
+	chainHeadChan      chan blockchain.ChainHeadEvent
+	chainHeadSub       event.Subscription
+	gov                governanceHelper
+	db                 database.DBManager
+	checkpointInterval uint64
+
+	// Internal data structures
+	accRewardCache *lru.ARCCache  // Cache (number uint64) -> (accReward *database.AccReward)
+	memoCache      *lru.ARCCache  // Cache (address Address) -> (memo.Burnt *big.Int)
+	quit           uint32         // Stop the goroutine in initial catchup stage
+	quitCh         chan struct{}  // Stop the goroutine in event subscription state
+	wg             sync.WaitGroup // background goroutine wait group for shutting down
+}
+
+// NewSupplyManager creates a new supply manager.
+// The TotalSupply data is stored every checkpointInterval blocks.
+func NewSupplyManager(chain blockChain, gov governanceHelper, db database.DBManager, checkpointInterval uint) *supplyManager {
+	accRewardCache, _ := lru.NewARC(supplyCacheSize)
+	memoCache, _ := lru.NewARC(10)
+
+	return &supplyManager{
+		chain:              chain,
+		chainHeadChan:      make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+		gov:                gov,
+		db:                 db,
+		checkpointInterval: uint64(checkpointInterval),
+		accRewardCache:     accRewardCache,
+		memoCache:          memoCache,
+		quitCh:             make(chan struct{}, 1), // make sure Stop() doesn't block if catchup() has exited before Stop()
+	}
+}
+
+func (sm *supplyManager) Start() {
+	sm.wg.Add(1)
```

Comment:
Shouldn't it check if one is already running, and just return if so?

Reply:
The Start() and Stop() are exclusively used in the [CN.Start() and CN.Stop()](https://github.com/klaytn/klaytn/blob/v1.12.1/node/cn/backend.go#L687-L726). Because we know exactly when they are called, and it's non-trivial to check that it's running, we don't check it. This is the same for [TxPool.Stop()](https://github.com/klaytn/klaytn/blob/v1.12.1/blockchain/tx_pool.go#L499) and [BlockChain.Stop()](https://github.com/klaytn/klaytn/blob/v1.12.1/blockchain/blockchain.go#L1075).